bomi/datastructure.py

```diff
@@ -108,12 +116,35 @@ def add_packet(self, packet: Packet):
         self.sensor_fp.write(",".join((str(v) for v in (packet.time, *readings))) + "\n")
 
         # Shift buffer when full, never changing buffer size
-        self.data[:-1] = self.data[1:]
-        self.data[-1] = readings
+        self._raw_data[:-1] = self._raw_data[1:]
+        self._raw_data[-1] = readings
         self.timestamp[:-1] = self.timestamp[1:]
         self.timestamp[-1] = packet.time
 
 
+class AveragedMultichannelBuffer(MultichannelBuffer):
+    DEFAULT_MOVING_AVERAGE_POINTS = 1024
+    """
+    The number of points for the moving average by default.
+    If the buffer is initialized with a size lower than this,
+    the number of points for the moving average will be the size.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.data = self._raw_data.copy()
+        self.moving_average_points = min(self.bufsize, self.DEFAULT_MOVING_AVERAGE_POINTS)
+
+    def add_packet(self, *args, **kwargs):
+        super().add_packet(*args, **kwargs)
+
+        moving_average_slice = self._raw_data[-self.moving_average_points:]
+        averages = tuple(moving_average_slice[col_name].mean() for col_name in moving_average_slice.dtype.names)
```

Comment:
Since `moving_average_slice` is a `recarray`, can you doing something like `moving_average_slice.mean(axis=-1)`?

Reply:
https://github.com/numpy/numpy/issues/24585